### PR TITLE
Feature - Training

### PR DESCRIPTION
**Pull request by ahmedheakl**
Feature - Training

## 📜 Summary
Here is the main juice of the repo! This branch is opt for creating the main trainers, training scripts, and entry points. 

###### Ticket(s): [#training](https://app.jira.com/training)

- 🔄 **Main changes in this PR**
    - Created a trainer interface for all models trainers
    - Implemented trainer for LLaMa model
    - Implemented training script with arguments parser.
    - Added evaluation script for inference. 

- ⚠️ **Breaking changes**
  - None

## 🚀 How to test this Pull Request

1. None

## 📝 Notes for future improvements

- Add testing script with evaluation metrics.

### CODE REVIEW

File: evaluate.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        help="Input assembly.",
+    )
+    parser.add_argument(
+        "--max-length",
+        type=int,
+        default=1024,
+        help="Maximum length of the output.",
+    )
+
+    args = parser.parse_args()
```
Comment (omar-abdelgawad) [inline evaluate.py:45]: Maybe making an argparse function that returns args is better.

File: train.py [hunk 2]
```diff
@@ -1,137 +1,64 @@
...
+        default="input",
+        help="Name of the input field in the dataset.",
+    )
+    parser.add_argument(
+        "--output_field_name",
+        type=str,
+        default="output",
+        help="Name of the output field in the dataset.",
+    )
+    args = parser.parse_args()
```
Comment (omar-abdelgawad) [inline train.py:49]: Same comment.

File: decompile/trainers/llama_trainer.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+        return (
+            "[INST] <<SYS>>\n"
+            + assembly_text
+            + "\n<</SYS>>\n\n"
+            + LLaMaOpt.instruction
+            + " [/INST] "
+        )
+
+    @staticmethod
+    def load_model(model_path: str, tokenizer_path: str):
```
Comment (omar-abdelgawad) [inline decompile/trainers/llama_trainer.py:239]: Return type?

File: decompile/trainers/llama_trainer.py [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+                        examples[input_field], examples[output_field]
+                    )
+                ]
+            }
+
+        train_dataset_mapped = self.train_dataset.map(mapper_handler, batched=True)
+        test_dataset_mapped = self.test_dataset.map(mapper_handler, batched=True)
+        return train_dataset_mapped, test_dataset_mapped
+
+    def train(self):
```
Comment (omar-abdelgawad) [inline decompile/trainers/llama_trainer.py:146]: Where is the return type of `Tuple[Any, Any]`

**Review by omar-abdelgawad — APPROVED** [review]
LGTM

File: evaluate.py (hunk 1 above)
Comment (ahmedheakl) [reply evaluate.py:45]: No need to do that now.

**Review by MohannedAhmed67 — APPROVED** [review]

**Pull request merged**